dbms/src/Common/ConcurrentIOQueue.h

```diff
@@ -0,0 +1,219 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/MPMCQueue.h>
+
+#include <deque>
+
+namespace DB
+{
+/** A simple thread-safe concurrent queue for io.
+  * Provide functions `nonBlockingPush` and `isFull` to support asynchronous writes.
+  * ```
+  * while (queue.isFull()) {}
+  * queue.nonBlockingPush(std::move(obj));
+  * ```
+  * There are additional overheads compared to using mpmcqueue directly
+  * - two locks
+  * - condition double check
+  * Since io queues are not performance sensitive, they are acceptable.
+  */
+template <typename T>
+class ConcurrentIOQueue
+{
+public:
+    explicit ConcurrentIOQueue(size_t capacity)
+        : mpmc_queue(std::max(1, capacity))
+    {}
+
+    ConcurrentIOQueue(
+        size_t capacity,
+        Int64 max_auxiliary_memory_usage_,
+        typename MPMCQueue<T>::ElementAuxiliaryMemoryUsageFunc && get_auxiliary_memory_usage_)
+        : mpmc_queue(std::max(1, capacity), max_auxiliary_memory_usage_, std::move(get_auxiliary_memory_usage_))
+    {}
+
+    /// blocking function.
+    /// Just like MPMCQueue::push.
+    MPMCQueueResult push(T && data)
+    {
+        return mpmc_queue.push(std::move(data));
+    }
+
+    /// Non-blocking function.
+    /// Besides all conditions mentioned at `push`, `nonBlockingPush` will still return OK if queue is `NORMAL` and full.
+    /// The obj that exceeds its capacity will be stored in remaining_objs and wait for the next pop/tryPop to trigger pushRemains.
+    MPMCQueueResult nonBlockingPush(T && data)
```

Comment:
```suggestion
    void push(T && data)
```
Since this function should only return OK

Reply:
Nope, it may return `FINISHED` or `CANCELLED`.

Reply:
> Nope, it may return `FINISHED` or `CANCELLED`.

Get it, should respect the mpmcqueue interface